### PR TITLE
Fix a string reference collision bug with AMF3

### PR DESCRIPTION
**Pull request by njoyce**

### CODE REVIEW

File: cpyamf/codec.pyx [hunk 1]
```diff
@@ -198,6 +198,26 @@ cdef class IndexedCollection(object):
...
+    There have been rare hash collisions within a single AMF payload causing
+    corrupt payloads.
+
+    Which strings cause collisions is dependent on the python runtime, each
+    platform might have a slightly different implementation which means that
+    testing is extremely difficult.
+    """
+
+    cdef inline object _ref(self, object obj):
+        print obj
```
Comment (johnlockwood-wf) [inline cpyamf/codec.pyx]: -1

File: cpyamf/codec.pyx [hunk 2]
```diff
@@ -198,6 +198,26 @@ cdef class IndexedCollection(object):
...
+        return obj
```
Comment (johnlockwood-wf) [inline cpyamf/codec.pyx]: very clever way to shorten the change of code needed.

File: cpyamf/codec.pyx [hunk 3]
```diff
@@ -198,6 +198,26 @@ cdef class IndexedCollection(object):
...
 
+cdef class ByteStringReferenceCollection(IndexedCollection):
+    """
+    There have been rare hash collisions within a single AMF payload causing
+    corrupt payloads.
+
+    Which strings cause collisions is dependent on the python runtime, each
+    platform might have a slightly different implementation which means that
+    testing is extremely difficult.
+    """
```
Comment (johnlockwood-wf) [inline cpyamf/codec.pyx]: you don't think the use of `cdef IndexedCollection s = self` in __richcmp__ will be a problem?

Comment (njoyce) [inline cpyamf/codec.pyx]: No - since they are accessing the same properties within IndexedCollection (and nothing in ByteStringReferenceCollection)

**Pull request merged**